Check for a lagging consumer

film42 opened the pull request:
This PR adds a heartbeat to the consumer thread. It also keeps the consumer thread ticking every `0.1` seconds. We will restart the consumer thread if a heartbeat is not received within 10 seconds.

I think the branch is close to ready, so if you can think of any concerns, please leave a comment.

Why do I think we should think about this? Recently we did an online restart of our rabbitmq cluster and we noticed active publisher getting stuck. We saw up to 2.5 minute delays to publish messages. Despite the publisher running slow, we had subscribers running without any problems. So the idea here is to kill any threads blocking the connection and retry. Plus, with this behavior, it makes future things like restarting the rabbitmq connection when the consumer is lagging a trivial thing.

NOTE: Depending on where the consumer thread is executing, it might drop the current batch of dequeued messages; however, this is existing behavior.

cc @mmmries @quixoten @abrandoned @brianstien @liveh2o @jjcarstens

film42 committed:
Add a heartbeat to the supervisor/ consumer

Consumer is lagging if all of the following are true:
1. We have not received a heartbeat from the consumer thread in more than 10 seconds.
2. A message has been pushed onto the queue since the last heartbeat message.
3. The supervisor has waited 5 "ticks" since (1) and (2) became true.

film42 committed:
Move the consumer heartbeat to the end of the consumer loop

film42 committed:
Simplify the consumer lagging logic by requiring a constant heartbeat

film42 committed:
Log an error message whenever we restart a lagging consumer

film42 committed:
Fix specs to work with real threads under jruby

film42 committed:
Prevent the consumer from overwhelming the supervisor with heartbeats

This supervisor clears them anyway, and if some thread-safe issue arise
where a clear happens as this message is being appended, we'll send a
new one on the next tick.

quixoten commented:
:+1:

abrandoned commented on lib/active_publisher/async/in_memory_adapter/async_queue.rb:
```diff
@@ -60,11 +61,28 @@ def size
+            logger.error "ActivePublisher consumer is lagging. Last heartbeat received #{seconds_since_last_heartbeat} seconds ago." if consumer_is_lagging
+
+            # Check to see if we should restart the consumer.
+            if !consumer.alive? || consumer_is_lagging
```
when this is lagging, is it safe to kill the consumer?

film42 replied:
Depending on where the consumer thread is executing, it might drop the current batch of dequeued messages. This is true of the consumer 💥 ing on some other random error.

abrandoned replied:
wondering if we would lose any messages really

film42 committed:
Make lagging consumer thread work on both mri and jruby

quixoten commented on lib/active_publisher/async/in_memory_adapter/consumer_thread.rb:
```diff
@@ -41,10 +42,19 @@ def make_channel
             loop do
+              # If the queue is empty, we should continue to send heartbeats to the consumer.
+              send_heartbeat
+              next sleep 0.1 if queue.empty?
```
this is optional, but if you want to remove the sleep we could merge in this: https://github.com/abrandoned/multi_op_queue/pull/1

then below

```ruby
current_messages = queue.pop_up_to(50, timeout: 0.1)
send_heartbeat
next unless current_messages
```

abrandoned replied:
like the idea of incorporating this into the `pop_up_to` call

film42 replied:
I agree! Very nice.

film42 committed:
Replace the heartbeat mechanism for a last_tick_at time on the consumer

This continues to clean things up quite a bit.

film42 committed:
Use new multi_op_queue timeout feature instead of sleeping

quixoten commented:
:shipit:

film42 committed:
Make the async publisher lag time configurable

film42 committed:
Wait until a successful publish to delete messages

Unless the message type is not a correct type, in which case we'll drop
the message

film42 commented:
@quixoten pushed a commit that makes the lag time configurable if you wouldn't mind lending your 👀 once more.

film42 committed:
Throw away any non ActivePublisher::Message instances first thing

quixoten commented:
:shipit:

film42 committed:
Drop messages that have rabbitmq precondition errors

film42 committed:
Add a test to verify that closing the channel crashes the consumer

film42 committed:
Merge pull request #39 from mxenabled/gt/only_delete_on_success

Wait for a successful publish before deleting messages

film42 commented:
Ok, I merged in the "only delete messages after successful publish" branch into here. Cool if I cut a pre-release of this branch?

film42 committed:
Bump version to 1.1.0.pre0

film42 commented:
Released `1.1.0.pre0`.

film42 commented:
Ok, this has been running smoothly for a week in production. Can I get an OK to ship?

abrandoned commented:
:shipit:
